File: c++/Minimum Cost to Hire K Workers.cpp
#include <vector>
#include <queue>
#include <algorithm>
#include <numeric>
#include <iostream>
#include <limits>

class Solution {
public:
    double mincostToHireWorkers(std::vector<int>& quality, std::vector<int>& wage, int k) {
        int n = quality.size();
        std::vector<std::pair<double, int>> ratioQuality;
        for (int i = 0; i < n; ++i) {
            ratioQuality.push_back({(double)wage[i] / quality[i], i});
        }
        std::sort(ratioQuality.begin(), ratioQuality.end());
        std::priority_queue<int> maxHeap;
        double minCost = std::numeric_limits<double>::max();
        double qualitySum = 0;
        for (auto [ratio, index] : ratioQuality) {
            qualitySum += quality[index];
            maxHeap.push(quality[index]);

            if (maxHeap.size() > k) {
                qualitySum -= maxHeap.top();
                maxHeap.pop();
            }
            if (maxHeap.size() == k) {
                minCost = std::min(minCost, ratio * qualitySum);
            }
        }
        return minCost;
    }
};
// test case
int main() {
    Solution solution;
    std::vector<int> quality1 = {10, 20, 5};
    std::vector<int> wage1 = {70, 50, 30};
    int k1 = 2;
    double result1 = solution.mincostToHireWorkers(quality1, wage1, k1);
    std::cout << "result pf test case: " << result1 << std::endl;
    return 0;
}

File: python/Minimum Cost to Hire K Workers.py
class Solution:
    def mincostToHireWorkers(self, quality: List[int], wage: List[int], k: int) -> float:
        n = len(quality)
        ratio_quality = [(wage[i] / quality[i], i) for i in range(n)]
        ratio_quality.sort()
        max_heap = []
        min_cost = float('inf')
        quality_sum = 0
        
        for ratio, index in ratio_quality:
            quality_sum += quality[index]
            heapq.heappush(max_heap, -quality[index])
            if len(max_heap) > k:
                quality_sum += heapq.heappop(max_heap)
            if len(max_heap) == k:
                min_cost = min(min_cost, ratio * quality_sum)

        return min_cost